DESIGN.md

```diff
@@ -0,0 +1,183 @@
+
+## Goal
+
+Gain insight in the license compatibility in software and its dependencies.
+
+## Purpose
+
+This tool should produce a report about the license compatibility for 1 on more pieces of software and their dependencies.
+
+## Background
+
+To use open source software a license needs to applied to it.
+Most software have dependencies which also have licenses.
+Not all licenses can be combined with each other.
+It is often difficult to find out the licenses of the dependencies and conflicts.
+
+## Personas
+
+At Netherlands eScience Center (NLeSC) there is need of more attention to the licensing issues. Personas in the center are
+
+-   program manager, someone responsible for software created in a program.
+-   research software engineer, someone writing software for a project in a program
+
+## User stories
+
+### Does this program have software that have potential license problems
+
+As a program manager at the NLeSC I would like have a tool that can find potential license problems in the software created in a program.
+
+### Does my piece of software have potential license problems
+
+As a research software engineer at the NLeSC I would like to find out if the dependencies of my piece of software has license violations.
+
+## Theoretical steps
+
+Regardless of whether we choose to use an existing tool/service or make something ourselves, the solution needs to have the following elements:
+
+1. input is a list of URLs to our repositories on GitHub, e.g. from Research Software Directory
+1. visit each, look in the root of the repository for a list of runtime (?) dependencies, for example from dependency files such as
+
+    1. requirements.txt
+    1. setup.cfg
+    1. setup.py
+    1. Pipfile
+    1. Pipfile.lock
+    1. pyproject.toml
+    1. environment.yml
+    1. yarn.lock
+    1. package.json
+    1. package-lock.json
+    1. DESCRIPTION
+
+    when repositories contain subdirectories with any of these dependency files, we need to do more nested/recursing evaluation
+
+1. for each of the project dependencies,
+    1. figure out whose copy is being used (e.g. PyPI, GitHub, crates.io, npm, maven, Anaconda cloud, bower, etc. more [here](https://en.wikipedia.org/wiki/List_of_software_package_management_systems#Application-level_package_managers))
+    1. identify the license as stated on the platform whose copy we're using
+1. yields a one-to-many graph (a tree) with relations ("If I have a GPL-3.0 here, can I have Apache-2.0 one level higher?")
+1. figure out if there is any license of a dependency that yields a conflict with the top level license (should be Apache-2.0) (based on which rules?)
+1. choose which result best represents the situation:
+    1. didn't find a dependency file
+    1. found a dependency file,
+        1. every name was resolvable
+            1. no conflicts
+            1. had conflicts (with list of conflicts found)
+        1. not every name was resolvable
+            1. needs human evaluation
+
```

Comment:
```suggestion
```

Reply:
I think we should keep this as it helped evaluate the existing tools